src/java/org/apache/cassandra/index/SecondaryIndexManager.java

```diff
@@ -1708,11 +1736,55 @@ public static void shutdownAndWait(long timeout, TimeUnit units) throws Interrup
         awaitTermination(timeout, units, asyncExecutor, blockingExecutor);
     }
 
+    /**
+     * Remove endpoints whose indexes are not queryable for the specified {@link Index.QueryPlan}.
+     *
+     * @param liveEndpoints current live endpoints where non-queryable endpoints will be removed
+     * @param keyspace to be queried
+     * @param indexQueryPlan index query plan used in the read command
+     * @param level consistency level of read command
+     */
+    public static <E extends Endpoints<E>> E filterForQuery(E liveEndpoints, Keyspace keyspace, Index.QueryPlan indexQueryPlan, ConsistencyLevel level)
```

Comment:
The complex of `filterForQuery()`, `propagateLocalIndexStatus()`, and `receivePeerIndexStatus()` seems essentially untested in this patch, unless I'm just having trouble finding those tests. My guess (memory is foggy here) is that any tests around this might be tied to on-disk index building, and therefore just aren't here yet. If that's the case, we could always make this patch simpler and remove as much as possible around these methods. (The `CQLTester` based stuff relies on `queryableIndexes`, not Gossip, right?) WDYT?

Reply:
I would prefer that we leave this functionality in because it is core to index stability even though we are in-memory only. What I have done is add a single distributed test that tests this functionality - `org.apache.cassandra.distributed.test.sai.IndexAvailabilityTest`.

I was, initially, going to add all the distributed tests on a single ticket when everything else had gone in but it makes sense now to add them as we go along to exercise this sort of functionality.